File: ListView/ShoppingCartItem/ShoppingCartItemVB/CartVisualItem.vb
Option Infer On

Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports Telerik.WinControls.UI


Public Class CartVisualItem
    Inherits SimpleListViewVisualItem

    Private buttonElement1 As RadButtonElement
    Private buttonElement2 As RadButtonElement
    Private titleElement As LightVisualElement
    Private descriptionElement As LightVisualElement
    Private priceElement As LightVisualElement
    Private amountElement As RadLabelElement
    Private stackLayout As StackLayoutElement
    Private stackLayout2 As StackLayoutElement
    Private stackLayout3 As StackLayoutElement

    Protected Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()
        'left part
        Me.stackLayout = New StackLayoutElement()
        Me.stackLayout.Orientation = Orientation.Vertical
        Me.stackLayout.ShouldHandleMouseInput = False
        Me.stackLayout.NotifyParentOnMouseInput = True
        Me.stackLayout.StretchHorizontally = True

        Me.titleElement = New LightVisualElement()
        Me.titleElement.Font = New Font("Segoe UI", 12, FontStyle.Regular)
        Me.titleElement.TextAlignment = ContentAlignment.MiddleLeft
        Me.titleElement.StretchHorizontally = True
        Me.titleElement.ShouldHandleMouseInput = False
        Me.titleElement.NotifyParentOnMouseInput = True

        Me.priceElement = New LightVisualElement()
        Me.priceElement.StretchHorizontally = True
        Me.priceElement.TextAlignment = ContentAlignment.MiddleLeft
        Me.priceElement.ShouldHandleMouseInput = False
        Me.priceElement.NotifyParentOnMouseInput = True

        Me.descriptionElement = New LightVisualElement()
        Me.descriptionElement.StretchHorizontally = True
        Me.descriptionElement.Font = New Font("Segoe UI", 7, FontStyle.Regular)
        Me.descriptionElement.TextAlignment = ContentAlignment.MiddleLeft
        Me.descriptionElement.ShouldHandleMouseInput = False
        Me.descriptionElement.NotifyParentOnMouseInput = True

        Me.stackLayout.Children.Add(Me.titleElement)
        Me.stackLayout.Children.Add(Me.priceElement)
        Me.stackLayout.Children.Add(Me.descriptionElement)

        'right part

        Me.stackLayout2 = New StackLayoutElement()
        Me.stackLayout2.Orientation = Orientation.Horizontal
        Me.stackLayout2.ShouldHandleMouseInput = False
        Me.stackLayout2.NotifyParentOnMouseInput = True
        Me.stackLayout2.StretchHorizontally = True
        Me.stackLayout2.StretchVertically = True

        Me.buttonElement1 = New RadButtonElement()
        Me.buttonElement1.Text = "-"
        Me.buttonElement1.Margin = New Padding(15)
        AddHandler Me.buttonElement1.Click, AddressOf ButtonElement1_Click
        Me.stackLayout2.Children.Add(Me.buttonElement1)

        Me.amountElement = New RadLabelElement()
        Me.amountElement.Margin = New Padding(15)

        Me.amountElement.TextAlignment = ContentAlignment.MiddleCenter
        Me.stackLayout2.Children.Add(amountElement)

        Me.buttonElement2 = New RadButtonElement()
        Me.buttonElement2.Margin = New Padding(15)
        Me.buttonElement2.Text = "+"
        AddHandler Me.buttonElement2.Click, AddressOf ButtonElement2_Click
        Me.stackLayout2.Children.Add(Me.buttonElement2)

        'combine
        Me.stackLayout3 = New StackLayoutElement()
        Me.stackLayout3.Orientation = Orientation.Horizontal
        Me.stackLayout3.ShouldHandleMouseInput = False
        Me.stackLayout3.NotifyParentOnMouseInput = True
        Me.stackLayout3.StretchHorizontally = True

        Me.stackLayout3.Children.Add(stackLayout)
        Me.stackLayout3.Children.Add(stackLayout2)
        Me.Children.Add(Me.stackLayout3)
    End Sub

    Private Sub ButtonElement2_Click(ByVal sender As Object, ByVal e As EventArgs)
        Dim data = TryCast(Me.Data.DataBoundItem, OrderItem)
        data.Amount += 1

    End Sub

    Private Sub ButtonElement1_Click(ByVal sender As Object, ByVal e As EventArgs)
        Dim data = TryCast(Me.Data.DataBoundItem, OrderItem)
        data.Amount -= 1

    End Sub

    Protected Overrides Sub SynchronizeProperties()
        MyBase.SynchronizeProperties()

        Me.Text = ""
        Dim data = TryCast(Me.Data.DataBoundItem, OrderItem)
        Me.titleElement.Text = data.Name
        Me.amountElement.Text = data.Amount.ToString()
        Me.priceElement.Text = data.Price.ToString("C")
        Me.descriptionElement.Text = data.QuantityPerUnit
    End Sub

    Protected Overrides ReadOnly Property ThemeEffectiveType() As Type
        Get
            Return GetType(SimpleListViewVisualItem)
        End Get
    End Property
End Class

File: ListView/ShoppingCartItem/ShoppingCartItemCS/CartVisualItem.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Telerik.WinControls.UI;

namespace _1165384
{
    public class CartVisualItem : SimpleListViewVisualItem
    {
        private RadButtonElement buttonElement1;
        private RadButtonElement buttonElement2;
        private LightVisualElement titleElement;
        private LightVisualElement descriptionElement;
        private LightVisualElement priceElement;
        private RadLabelElement amountElement;
        private StackLayoutElement stackLayout;
        private StackLayoutElement stackLayout2;
        private StackLayoutElement stackLayout3;

        protected override void CreateChildElements()
        {
            base.CreateChildElements();
            //left part
            this.stackLayout = new StackLayoutElement();
            this.stackLayout.Orientation = Orientation.Vertical;
            this.stackLayout.ShouldHandleMouseInput = false;
            this.stackLayout.NotifyParentOnMouseInput = true;
            this.stackLayout.StretchHorizontally = true;

            this.titleElement = new LightVisualElement();
            this.titleElement.Font = new Font("Segoe UI", 12, FontStyle.Regular);
            this.titleElement.TextAlignment = ContentAlignment.MiddleLeft;
            this.titleElement.StretchHorizontally = true;
            this.titleElement.ShouldHandleMouseInput = false;
            this.titleElement.NotifyParentOnMouseInput = true;

            this.priceElement = new LightVisualElement();
            this.priceElement.StretchHorizontally = true;
            this.priceElement.TextAlignment = ContentAlignment.MiddleLeft;
            this.priceElement.ShouldHandleMouseInput = false;
            this.priceElement.NotifyParentOnMouseInput = true;

            this.descriptionElement = new LightVisualElement();
            this.descriptionElement.StretchHorizontally = true;
            this.descriptionElement.Font = new Font("Segoe UI", 7, FontStyle.Regular);
            this.descriptionElement.TextAlignment = ContentAlignment.MiddleLeft;
            this.descriptionElement.ShouldHandleMouseInput = false;
            this.descriptionElement.NotifyParentOnMouseInput = true;

            this.stackLayout.Children.Add(this.titleElement);
            this.stackLayout.Children.Add(this.priceElement);
            this.stackLayout.Children.Add(this.descriptionElement);

            //right part

            this.stackLayout2 = new StackLayoutElement();
            this.stackLayout2.Orientation = Orientation.Horizontal;
            this.stackLayout2.ShouldHandleMouseInput = false;
            this.stackLayout2.NotifyParentOnMouseInput = true;
            this.stackLayout2.StretchHorizontally = true;
            this.stackLayout2.StretchVertically = true;

            this.buttonElement1 = new RadButtonElement();
            this.buttonElement1.Text = "-";
            this.buttonElement1.Margin = new Padding(15);
            this.buttonElement1.Click += ButtonElement1_Click;
            this.stackLayout2.Children.Add(this.buttonElement1);

            this.amountElement = new RadLabelElement();
            this.amountElement.Margin = new Padding(15);

            this.amountElement.TextAlignment = ContentAlignment.MiddleCenter;
            this.stackLayout2.Children.Add(amountElement);

            this.buttonElement2 = new RadButtonElement();
            this.buttonElement2.Margin = new Padding(15);
            this.buttonElement2.Text = "+";
            this.buttonElement2.Click += ButtonElement2_Click;
            this.stackLayout2.Children.Add(this.buttonElement2);

            //combine
            this.stackLayout3 = new StackLayoutElement();
            this.stackLayout3.Orientation = Orientation.Horizontal;
            this.stackLayout3.ShouldHandleMouseInput = false;
            this.stackLayout3.NotifyParentOnMouseInput = true;
            this.stackLayout3.StretchHorizontally = true;

            this.stackLayout3.Children.Add(stackLayout);
            this.stackLayout3.Children.Add(stackLayout2);
            this.Children.Add(this.stackLayout3);
        }

        private void ButtonElement2_Click(object sender, EventArgs e)
        {
            var data = this.Data.DataBoundItem as OrderItem;
            data.Amount++;

        }

        private void ButtonElement1_Click(object sender, EventArgs e)
        {
            var data = this.Data.DataBoundItem as OrderItem;
            data.Amount--;

        }

        protected override void SynchronizeProperties()
        {
            base.SynchronizeProperties();

            this.Text = "";
            var data = this.Data.DataBoundItem as OrderItem;
            this.titleElement.Text = data.Name;
            this.amountElement.Text = data.Amount.ToString();
            this.priceElement.Text = data.Price.ToString("C");
            this.descriptionElement.Text = data.QuantityPerUnit;
        }

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(SimpleListViewVisualItem);
            }
        }
    }
}
